File: dotnet/Core/GameX/Formats/IUnknown.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Numerics;

namespace GameX.Formats.Unknown;

#region IUnknownBone

public interface IUnknownBone {
    string Name { get; }
    Matrix4x4 WorldToBone { get; } // 4x3 matrix
    Matrix4x4 BoneToWorld { get; } // 4x3 matrix of world translations/rotations of the bones.
}

#endregion

#region IUnknownFileModel

public interface IUnknownFileModel : IUnknownFileObject {
    IEnumerable<IUnknownModel> Models { get; }
    IEnumerable<UnknownMesh> Meshes { get; }
    IEnumerable<IUnknownMaterial> Materials { get; }
    IEnumerable<IUnknownProxy> Proxies { get; }
    IUnknownSkin SkinningInfo { get; }
    IEnumerable<string> RootNodes { get; }
}

#endregion

#region IUnknownFileObject

public interface IUnknownFileObject {
    public struct Source {
        public string Author;
        public string SourceFile;
    }

    string Name { get; }
    string Path { get; }
    IEnumerable<Source> Sources { get; }
}

#endregion

#region IUnknownMaterial

public interface IUnknownMaterial {
    string Name { get; }
    Vector3? Diffuse { get; } // Color:RGB
    Vector3? Specular { get; } // Color:RGB
    Vector3? Emissive { get; } // Color:RGB
    float Shininess { get; }
    float Opacity { get; }
    IEnumerable<IUnknownTexture> Textures { get; }
}

#endregion

#region IUnknownModel

public interface IUnknownModel {
    string Path { get; }
}

#endregion

#region IUnknownProxy

public interface IUnknownProxy {
    public struct Proxy {
        public Vector3[] Vertexs;
        public int[] Indexs;
    }

    Proxy[] PhysicalProxys { get; }
}

#endregion

#region IUnknownSkin

public interface IUnknownSkin {
    public struct BoneMap {
        public int[] BoneIndex;
        public int[] Weight; // Byte / 256?
    }

    public struct IntVertex {
        public Vector3 Obsolete0;
        public Vector3 Position;
        public Vector3 Obsolete2;
        public ushort[] BoneIDs; // 4 bone IDs
        public float[] Weights; // Should be 4 of these
        public object Color;
    }

    bool HasSkinningInfo { get; }
    ICollection<IUnknownBone> CompiledBones { get; }
    IntVertex[] IntVertexs { get; }
    BoneMap[] BoneMaps { get; }
    ushort[] Ext2IntMaps { get; }
}

#endregion

#region IUnknownTexture

public interface IUnknownTexture {
    [Flags]
    public enum Map {
        Diffuse = 1 << 0,
        Bumpmap = 1 << 1,
        Specular = 1 << 2,
        Environment = 1 << 3,
        Decal = 1 << 4,
        SubSurface = 1 << 5,
        Opacity = 1 << 6,
        Detail = 1 << 7,
        Heightmap = 1 << 8,
        BlendDetail = 1 << 9,
        Custom = 1 << 10,
    }

    string Path { get; }
    Map Maps { get; }
}

#endregion

#region IUnknownUnknown

//public enum ChunkType
//{
//    Node = 1,
//    Mesh,
//    Helper,
//}

//IGenericNode Root { get; }
//IEnumerable<IGenericNode> Nodes { get; }

//public interface IGenericNode
//{
//    string Name { get; }
//    ChunkType Type { get; }
//    IGenericNode Parent { get; }
//    IChunk Object { get; }
//    public IEnumerable<IGenericNode> Children { get; set; }
//}

//public interface IChunk
//{
//    ChunkType Type { get; }
//}

//public interface IChunkMesh : IChunk
//{
//    int Id { get; set; }
//    int MeshSubsets { get; set; }
//    int VerticesData { get; set; }
//    int VertsUVsData { get; set; }
//}

#endregion

#region UnknownFileWriter

public abstract class UnknownFileWriter {
    public static readonly Dictionary<string, Func<IUnknownFileModel, UnknownFileWriter>> Factories = new Dictionary<string, Func<IUnknownFileModel, UnknownFileWriter>>(StringComparer.OrdinalIgnoreCase);

    // ARGS
    public DirectoryInfo DataDir = null;
    public const bool NoConflicts = false;
    public const bool TiffTextures = false;
    public const bool SkipShieldNodes = false;
    public const bool SkipStreamNodes = false;
    public const bool GroupMeshes = true;
    public const bool Smooth = true;

    public IUnknownFileModel File { get; internal set; }

    public UnknownFileWriter(IUnknownFileModel file) => File = file;

    public abstract void Write(string outputDir = null, bool preservePath = true);

    protected FileInfo GetFileInfo(string extension, string outputDir = null, bool preservePath = true) {
        var fileName = $"temp.{extension}";
        // Empty output directory means place alongside original models If you want relative path, use "."
        if (string.IsNullOrWhiteSpace(outputDir)) fileName = Path.Combine(new FileInfo(File.Path).DirectoryName, $"{Path.GetFileNameWithoutExtension(File.Path)}{(NoConflicts ? "_out" : string.Empty)}.{extension}");
        else {
            // If we have an output directory
            var preserveDir = preservePath ? Path.GetDirectoryName(File.Path) : string.Empty;
            // Remove drive letter if necessary
            if (!string.IsNullOrWhiteSpace(preserveDir) && !string.IsNullOrWhiteSpace(Path.GetPathRoot(preserveDir))) preserveDir = preserveDir.Replace(Path.GetPathRoot(preserveDir), string.Empty);
            fileName = Path.Combine(outputDir, preserveDir, Path.ChangeExtension(Path.GetFileNameWithoutExtension(File.Path), extension));
        }
        return new FileInfo(fileName);
    }

    public static UnknownFileWriter Factory(string name, IUnknownFileModel model) => Factories.TryGetValue(name, out var factory)
        ? factory(model)
        : throw new ArgumentOutOfRangeException(nameof(name), name);
}

#endregion

#region UnknownMesh

public class UnknownMesh {
    [Flags]
    public enum Effect {
        ScaleOffset = 0x1
    }

    public struct Subset {
        public Range Vertexs;
        public Range Indexs;
        public int MatId;
        //public float Radius;
        //public Vector3 Center;
    }

    public ref struct SubsetMesh {
        public Span<Vector3> Vertexs;
        public Span<int> Indexs;
        public Span<Vector3> Normals;
        public Span<Vector2> UVs;
    }

    public string Name;
    public Vector3 MinBound;
    public Vector3 MaxBound;
    public Subset[] Subsets;
    public Vector3[] Vertexs;
    public Vector2[] UVs;
    public Vector3[] Normals;
    public int[] Indexs;
    public Effect Effects;
    public (Vector3 scale, Vector3 offset) ScaleOffset3;
    public (Vector4 scale, Vector4 offset) ScaleOffset4;

    public SubsetMesh this[Subset i] => new() {
        Vertexs = Vertexs.AsSpan(i.Vertexs),
        UVs = UVs.AsSpan(i.Vertexs),
        Normals = Normals != null ? Normals.AsSpan(i.Vertexs) : null,
        Indexs = Indexs.AsSpan(i.Indexs),
    };

    public Vector3 GetTransform(Vector3 vertex) => vertex;
}

#endregion

File: rust/core/gamex/src/formats/i_unknown.rs
// PORT-SOURCE: Core/GameX/Formats/IUnknown.cs
// PORT-SHA: d536d2fcd7b2d724
// PORT-STATUS: done
//
// The abstract 3D-model interfaces every family's model reader implements:
// `IUnknownFileModel`, `IUnknownMaterial`, `IUnknownSkin`, `IUnknownBone` and
// friends. Pure declarations — 166 lines with no bodies.
//
// ===================== THREE C#-SIDE OBSERVATIONS ========================
//
//   1. **`IUnknownBone` declares two `Matrix4x4` properties that the comments
//      say are 4x3.** `WorldToBone` and `BoneToWorld` are both annotated
//      `// 4x3 matrix`, so the fourth row is padding every implementor has to
//      know to ignore — and nothing enforces that it is identity. `Mat4` is
//      kept here (glam has no 4x3) with the convention documented, because
//      changing the storage would change every family's reader.
//
//   2. **`IUnknownSkin.BoneMap.Weight` is `int[]`, commented `// Byte / 256?`.**
//      The comment is a question mark in shipped code: nothing else says
//      whether these are 0..255 or 0..1 scaled, and the two differ by a factor
//      of 256 in every skinned vertex. Left as `i32` with the ambiguity
//      recorded — **this needs someone who knows the format.**
//
//   3. **`IntVertex` has fields named `Obsolete0` and `Obsolete2`.** Two of its
//      three `Vector3`s are declared obsolete by name but still read and
//      stored, so every skinned vertex carries 24 bytes nothing consumes.
//
// Ported as traits. Rust has no property syntax, so getters become methods;
// `IEnumerable<T>` becomes a slice where the data is owned and an iterator
// where it is computed — noted per method.

use glam::{Mat4, Vec3};

/// C# `IUnknownFileObject.Source`.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Source {
    pub author: Option<String>,
    pub source_file: Option<String>,
}

/// C# `IUnknownFileObject`.
pub trait UnknownFileObject {
    fn name(&self) -> &str;
    fn path(&self) -> &str;
    fn sources(&self) -> &[Source];
}

/// C# `IUnknownBone`.
pub trait UnknownBone {
    fn name(&self) -> &str;
    /// C# `WorldToBone` — commented "4x3 matrix"; the fourth row is padding.
    fn world_to_bone(&self) -> Mat4;
    /// C# `BoneToWorld` — likewise 4x3.
    fn bone_to_world(&self) -> Mat4;
}

/// C# `IUnknownTexture`.
pub trait UnknownTexture {
    fn path(&self) -> &str;
}

/// C# `IUnknownMaterial`.
///
/// The three colours are `Vector3?` in the C# — RGB with no alpha, optional.
pub trait UnknownMaterial {
    fn name(&self) -> &str;
    fn diffuse(&self) -> Option<Vec3>;
    fn specular(&self) -> Option<Vec3>;
    fn emissive(&self) -> Option<Vec3>;
    fn shininess(&self) -> f32;
    fn opacity(&self) -> f32;
    fn textures(&self) -> Vec<&dyn UnknownTexture>;
}

/// C# `IUnknownModel`.
pub trait UnknownModel {
    fn path(&self) -> &str;
}

/// C# `IUnknownProxy.Proxy`.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Proxy {
    pub vertexs: Vec<Vec3>,
    pub indexs: Vec<i32>,
}

impl Proxy {
    /// Whether every index addresses a vertex that exists.
    ///
    /// No C# equivalent — the arrays are read from a file and handed to the
    /// renderer unchecked.
    pub fn indices_in_range(&self) -> bool {
        let n = self.vertexs.len() as i32;
        self.indexs.iter().all(|&i| i >= 0 && i < n)
    }
}

/// C# `IUnknownProxy`.
pub trait UnknownProxy {
    fn physical_proxys(&self) -> &[Proxy];
}

/// C# `IUnknownSkin.BoneMap`.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct BoneMap {
    pub bone_index: Vec<i32>,
    /// C# comment: `// Byte / 256?` — see observation 2. Scale is unresolved.
    pub weight: Vec<i32>,
}

/// C# `IUnknownSkin.IntVertex`.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct IntVertex {
    /// C# `Obsolete0` — read and stored, consumed by nothing.
    pub obsolete0: Vec3,
    pub position: Vec3,
    /// C# `Obsolete2` — likewise.
    pub obsolete2: Vec3,
    /// C# comment: "4 bone IDs".
    pub bone_ids: [u16; 4],
    /// C# comment: "Should be 4 of these" — a `float[]` with an expectation
    /// rather than a fixed size, so nothing rejects a vertex with three.
    pub weights: [f32; 4],
    pub color: Option<u32>,
}

impl IntVertex {
    /// Whether the four weights sum to 1 within tolerance.
    ///
    /// No C# equivalent, and worth having: a mis-scaled weight array (see
    /// observation 2) shows up here rather than as silently wrong deformation.
    pub fn weights_normalised(&self, tol: f32) -> bool {
        (self.weights.iter().sum::<f32>() - 1.0).abs() <= tol
    }
}

/// C# `IUnknownSkin`.
pub trait UnknownSkin {
    fn has_skinning_info(&self) -> bool;
    fn compiled_bones(&self) -> Vec<&dyn UnknownBone>;
    fn int_vertexs(&self) -> &[IntVertex];
}

/// C# `IUnknownFileModel`.
pub trait UnknownFileModel: UnknownFileObject {
    fn models(&self) -> Vec<&dyn UnknownModel>;
    fn materials(&self) -> Vec<&dyn UnknownMaterial>;
    fn proxies(&self) -> Vec<&dyn UnknownProxy>;
    fn skinning_info(&self) -> Option<&dyn UnknownSkin>;
    fn root_nodes(&self) -> Vec<&str>;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn proxy_indices_are_range_checked() {
        let p = Proxy { vertexs: vec![Vec3::ZERO; 3], indexs: vec![0, 1, 2] };
        assert!(p.indices_in_range());
        let bad = Proxy { vertexs: vec![Vec3::ZERO; 3], indexs: vec![0, 3] };
        assert!(!bad.indices_in_range(), "3 is past the end of 3 vertices");
        let neg = Proxy { vertexs: vec![Vec3::ZERO], indexs: vec![-1] };
        assert!(!neg.indices_in_range());
    }

    #[test]
    fn an_empty_proxy_is_trivially_in_range() {
        assert!(Proxy::default().indices_in_range());
    }

    #[test]
    fn normalised_weights_are_detectable() {
        let v = IntVertex { weights: [0.5, 0.25, 0.25, 0.0], ..Default::default() };
        assert!(v.weights_normalised(1e-6));
        // The `Byte / 256?` ambiguity looks like this if read at the wrong scale.
        let unscaled = IntVertex { weights: [128.0, 64.0, 64.0, 0.0], ..Default::default() };
        assert!(!unscaled.weights_normalised(1e-6));
        assert!(
            (unscaled.weights.iter().sum::<f32>() / 256.0 - 1.0).abs() < 1e-6,
            "and /256 is what makes it normalise, which is the open question"
        );
    }

    #[test]
    fn bone_ids_and_weights_are_fixed_at_four() {
        // The C# uses ushort[] and float[] with the count only in a comment.
        let v = IntVertex::default();
        assert_eq!(v.bone_ids.len(), 4);
        assert_eq!(v.weights.len(), 4);
    }

    #[test]
    fn obsolete_fields_are_present_as_the_c_sharp_has_them() {
        // Named obsolete, still read, consumed by nothing - 24 bytes per vertex.
        let v = IntVertex::default();
        assert_eq!(v.obsolete0, Vec3::ZERO);
        assert_eq!(v.obsolete2, Vec3::ZERO);
    }
}
